### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by rmzse**
## Pivotal Tracker story 

https://www.pivotaltracker.com/story/show/137254137

## Changes proposed in this pull request:

This feature creates the sign up functionality to enable users to sign up and create a user account in the system.

## What I have learned working on this feature:

• recapping TDD and acceptance testing
• installing and setting up Devise
• validating models and factories

#### Ready for review!

### CODE REVIEW

**Comment by tochman** [comment]
There are merge conflicts here. When will this be worked on?

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  validates :encrypted_password, presence: true
+  validates :reset_password_token, presence: true
+  validates :reset_password_sent_at, presence: true
+  validates :remember_created_at, presence: true
+  validates :sign_in_count, presence: true
+  validates :current_sign_in_at, presence: true
+  validates :last_sign_in_at, presence: true
+  validates :current_sign_in_ip, presence: true
+  validates :last_sign_in_ip, presence: true
```
Comment (tochman) [inline app/models/user.rb]: None of this validations are needed. Please stick to the Devise documentation

File: features/step_definitions/assertion_steps.rb [hunk 2]
```diff
@@ -12,4 +12,12 @@
...
\ No newline at end of file
+end
+
+Then(/^I should be on the signup page$/) do
+  expect(current_path).to eq new_user_registration_path
+end
+
+# And(/^a user with "([^"]*)" and "([^"]*)" should be added$/) do |sign_up_email, sign_up_password|
+#   Factory.create(:user, sign_up_email, sign_up_password)
+# end
```
Comment (tochman) [inline features/step_definitions/assertion_steps.rb]: Get rid of comments please

File: features/step_definitions/sign_up_steps.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+	fill_in "Email", with: :email
+end
+
+And(/^I enter my password in the "([^"]*)" field$/) do |field_name|
+	fill_in "Password", with: :password
+end
+
+And(/^I click the "([^"]*)" button$/) do |button|
+	click_button button
+end
```
Comment (tochman) [inline features/step_definitions/sign_up_steps.rb:7]: Please fix settings on the IDE/editor to properly format code.

File: spec/factories/users.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    password 'password'
+    encrypted_password 'encrypted_password'
+    reset_password_token 'reset_password_token'
+    reset_password_sent_at '2017-01-10 16:32:43'
+    remember_created_at '2017-01-08 16:32:43'
+    sign_in_count 1
+    current_sign_in_at '2017-01-15 16:32:43'
+    last_sign_in_at '2017-01-14 16:32:43'
+    current_sign_in_ip '127.0.0.0'
+    last_sign_in_ip '127.0.0.0'
```
Comment (tochman) [inline spec/factories/users.rb]: Most of these settings are not needed. Please see comment about validations

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      it { is_expected.to validate_presence_of :encrypted_password }
+      it { is_expected.to validate_presence_of :reset_password_token }
+      it { is_expected.to validate_presence_of :reset_password_sent_at }
+      it { is_expected.to validate_presence_of :remember_created_at }
+      it { is_expected.to validate_presence_of :sign_in_count }
+      it { is_expected.to validate_presence_of :current_sign_in_at }
+      it { is_expected.to validate_presence_of :last_sign_in_at }
+      it { is_expected.to validate_presence_of :current_sign_in_ip }
+      it { is_expected.to validate_presence_of :last_sign_in_ip }
+    end
```
Comment (tochman) [inline spec/models/user_spec.rb:8]: No need to validate most of these attributes. Devise takes care of most of that.

**Review by tochman — CHANGES_REQUESTED** [review]

**Review by diraulo — COMMENTED** [review]
@rmzse @kriizjones Are you guys overriding any of the Devise views with your own content/styles? If not then all the Devise generated views are not necessary and can be deleted.

Also, I'm afraid this PR is seriously growing out of scope. Creating a PR template really has nothing to do with User Signup for example.

File: README.md [hunk 6]
```diff
@@ -56,7 +57,7 @@ Give an example
 
 ## Deployment
 
-Add additional notes about how to deploy this on a live system
+We are using Travis for continous integration and Heroku for deployment.
```
Comment (tochman) [inline README.md]: Why on earth are you including the README updates in this PR? Please remove this.

**Review by tochman — CHANGES_REQUESTED** [review]

File: features/step_definitions/sign_up_steps.rb (hunk 3 above)
Comment (rmzse) [reply features/step_definitions/sign_up_steps.rb:7]: @tochman we are not finding this err, as Rubymine says it's already fixed. We'll try one more thing and see if we succeed.

image

File: features/sign_up.feature [hunk 7]
```diff
@@ -0,0 +1,15 @@
+Feature: Sign Up
+  As a visitor,
+  In order to buy and sell tasty food,
+  I need to be able to get system access by creating an account
+
+  Scenario: Successfully Signed Up
+    When I am on the "landing" page
+    And I click the "Sign up" link
+    Then I should be on the "Sign up" page
```
Comment (tochman) [inline features/sign_up.feature]: We spoke about assertions like this in our last session. Please remove this.

**Review by tochman — CHANGES_REQUESTED** [review]
Much better, almost there.

File: features/sign_up.feature (hunk 7 above)
Comment (rmzse) [reply features/sign_up.feature]: Yes, we did! 
Sorry, forgot this one in the midst of everything else...

Comment (rmzse) [reply features/sign_up.feature]: @tochman Code review please! :)

Comment (tochman) [reply features/sign_up.feature]: Why is this step still here?

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+    describe 'DB table' do
+      it { is_expected.to have_db_column :id }
+      it { is_expected.to have_db_column :email }
+      it { is_expected.to have_db_column :encrypted_password }
+    end
+
+
```
Comment (tochman) [inline spec/models/user_spec.rb]: Remove empty lines please.

**Comment by tochman** [comment]
Still WIP? Please fix conflicts.

**Review by tochman — APPROVED** [review]

**Pull request merged**